### PR TITLE
feat: add push flag to support pushing to registry directly

### LINKED ISSUE #293: [REQ] Support pushing to registry directly
### What kind of request is this?

Improvement of existing experience

### What is your request or suggestion?

It would be nice to push to the image registry directly. 
Like : 
```
copa patch -i docker.io/library/alpine:3.16.4 -r patch.json -p ttl.sh/alpinepatched:1h
```

One side-effect of this would be that copa wouldn't rely on Docker anymore if there is a remote buildkit setup as I mentioned here. https://github.com/project-copacetic/copacetic/issues/285#issuecomment-1708239927

### PR DESCRIPTION
**Pull request by RealHarshThakur**
This PR enables direct push to registry from buildkit:
```
copa patch -i docker.io/library/alpine:3.16.4 -r patch.json -p ttl.sh/alpinepatched:1h
```

One side-effect of this would be that copa wouldn't rely on Docker anymore if there is a remote buildkit setup as I mentioned here. https://github.com/project-copacetic/copacetic/issues/285#issuecomment-1708239927

Closes #293

### CODE REVIEW

**Comment by sozercan** [comment]
I am wondering if we should align with buildx CLI here. Ideally, we don't want to break existing users so what I was thinking was to use `--tag` to accept `registy/repo:tag` and `--push` can be a bool. to remain backwards compat, we can parse if `--tag` includes `:` and `/`. If it includes both `:` and `/`, retag the entire image. If it doesn't include both `:` and `/`, only replace the tag.

`-i foo.io/repo/image:tag -t -patched` -> foo.io/repo/image:tag-patched (what we have today)
`-i foo.io/repo/image:tag -t bar.io/repo/image:tag` ->  bar.io/repo/image:tag
`-i foo.io/repo/image:tag -t bar.io/repo/image` -> invalid (must provide tag)
`-i foo.io/repo/image:tag -t repo/image:tag` -> invalid (must provide registry)

@jeremyrickard @cpuguy83 wdyt?

**Comment by jeremyrickard** [comment]
> I am wondering if we should align with buildx CLI here. Ideally, we don't want to break existing users so what I was thinking was to use `--tag` to accept `registy/repo:tag` and `--push` can be a bool. to remain backwards compat, we can parse if `--tag` includes `:` and `/`. If it includes both `:` and `/`, retag the entire image. If it doesn't include both `:` and `/`, only replace the tag.
> 
> `-i foo.io/repo/image:tag -t -patched` -> foo.io/repo/image:tag-patched (what we have today) `-i foo.io/repo/image:tag -t bar.io/repo/image:tag` -> bar.io/repo/image:tag `-i foo.io/repo/image:tag -t bar.io/repo/image` -> invalid (must provide tag) `-i foo.io/repo/image:tag -t repo/image:tag` -> invalid (must provide registry)
> 
> @jeremyrickard @cpuguy83 wdyt?

I like the `-i foo.io/repo/image:tag -t bar.io/repo/image:tag` syntax. 

I also think that adding a --push flag would be useful, especially if it didn't require docker locally.  Not the same scenario, but the experience with`ko` and pushing images w/o docker is super useful.

**Comment by RealHarshThakur** [comment]
cool, happy to go with what you folks suggested. 

Although, I am not sure about this bit:
```
-i foo.io/repo/image:tag -t repo/image:tag -> invalid (must provide registry)
```

This might still be what the user wants, partially due to how popular dockerhub is: `username/image:tag` is valid over there. Also, other such registries like `ttl.sh` also don't mind that format.

**Comment by RealHarshThakur** [comment]
I've pushed a commit to get most of what you folks suggested. Added couple test cases that shows we'll be backwards-compatible and a new case for specifying a different registry. I've tried to stay unopinionated about destination registry conventions as copa will throw an error during push if buildkit fails to push anyway. Happy to take your suggestions tho

**Comment by RealHarshThakur** [comment]
I think setting up the mock buildkit API  in unit tests failed, is it acting flaky?

**Comment by sozercan** [comment]
> This might still be what the user wants, partially due to how popular dockerhub is: username/image:tag is valid over there. Also, other such registries like ttl.sh also don't mind that format.

this is being consistent with `-i` since you must provide `docker.io` there

**Comment by RealHarshThakur** [comment]
> this is being consistent with -i since you must provide docker.io there

cool, I've added the registry check  @sozercan

**Comment by RealHarshThakur** [comment]
@sozercan @jeremyrickard don't mean to badger, is there anything else you'd like me to change on this PR?

**Comment by salaxander** [comment]
@RealHarshThakur sorry for the delay - @sozercan has been on vacation for the past couple days. I'll see if I can get you a review asap :)

File: pkg/patch/patch_test.go [hunk 1]
```diff
@@ -49,3 +49,58 @@ func TestRemoveIfNotDebug(t *testing.T) {
...
+			name:       "tag passed with value",
+			image:      "docker.io/library/nginx:1.21.3",
+			patchedTag: "custom",
+			want:       "docker.io/library/nginx:custom",
+			wantErr:    false,
+		},
+		{
+			name:       "tag passed but without registry or repo",
+			image:      "docker.io/library/nginx:1.21.3",
+			patchedTag: "my.registry/nginx:1.21.3-patched",
```
Comment (sozercan) [inline pkg/patch/patch_test.go:79]: this should be valid (something like `myregistry.azurecr.io/nginx:1.21.3-patched`)

File: pkg/patch/patch.go [hunk 2]
```diff
@@ -160,3 +147,45 @@ func patchWithContext(ctx context.Context, image, reportFile, patchedTag, workin
...
+		} else {
+			patchedTag = fmt.Sprintf("%s-%s", tag, defaultPatchedTagSuffix)
+		}
+	}
+
+	slashCount := strings.Count(patchedTag, "/")
+	if slashCount > 0 {
+		if slashCount < 2 {
+			err := fmt.Errorf("invalid tag %s, must be in the form <registry>/<image>:<tag>", patchedTag)
+			log.Error(err)
```
Comment (sozercan) [inline pkg/patch/patch.go:181]: i think this is duplicate, since return is logged already?

**Review by sozercan — CHANGES_REQUESTED** [review]
Sorry about the delay! Please see comments

**Comment by sozercan** [comment]
ping @RealHarshThakur. we are getting close to cutting v0.5.0 release if you want this to be included

**Comment by RealHarshThakur** [comment]
I'll try but most likely,  I'll be able to pick it up by end of the month

**Comment by RealHarshThakur** [comment]
@sozercan  sorry for the delay, I've made the changes you've requested.

**Comment by sozercan** [comment]
Closing due to inactivity. Please feel free to re-open if there's ongoing work. Thanks!

**Pull request closed without merging**